tools/code_cache/cache_builder.cc

```diff
@@ -0,0 +1,142 @@
+#include "cache_builder.h"
+#include "node_native_module.h"
+#include "util.h"
+
+#include <iostream>
+#include <map>
+#include <sstream>
+#include <vector>
+
+namespace node {
+namespace native_module {
+
+using v8::Context;
+using v8::Function;
+using v8::Isolate;
+using v8::Local;
+using v8::MaybeLocal;
+using v8::ScriptCompiler;
+
+static std::string GetDefName(const std::string& id) {
+  char buf[64] = {0};
+  size_t size = id.size();
+  CHECK_LT(size, sizeof(buf));
+  for (size_t i = 0; i < size; ++i) {
+    char ch = id[i];
+    buf[i] = (ch == '-' || ch == '/') ? '_' : ch;
+  }
+  return buf;
+}
+
+static std::string FormatSize(size_t size) {
+  char buf[64] = {0};
+  if (size < 1024) {
+    snprintf(buf, sizeof(buf), "%.2fB", static_cast<double>(size));
+  } else if (size < 1024 * 1024) {
+    snprintf(buf, sizeof(buf), "%.2fKB", static_cast<double>(size / 1024));
+  } else {
+    snprintf(
+        buf, sizeof(buf), "%.2fMB", static_cast<double>(size / 1024 / 1024));
+  }
+  return buf;
+}
+
+static std::string GetDefinition(const std::string& id,
+                                 size_t size,
+                                 const uint8_t* data) {
+  std::stringstream ss;
+  ss << "static const uint8_t " << GetDefName(id) << "[] = {\n";
+  for (size_t i = 0; i < size; ++i) {
+    uint8_t ch = data[i];
+    ss << std::to_string(ch) << (i == size - 1 ? '\n' : ',');
+  }
+  ss << "};";
+  return ss.str();
+}
+
+static void GetInitializer(const std::string& id, std::stringstream& ss) {
+  std::string def_name = GetDefName(id);
+  ss << "  code_cache_.emplace(\n";
+  ss << "    \"" << id << "\",\n";
+  ss << "    std::make_unique<v8::ScriptCompiler::CachedData>(\n";
+  ss << "      " << def_name << ",\n";
+  ss << "      static_cast<int>(arraysize(" << def_name << ")), policy\n";
+  ss << "    )\n";
+  ss << "  );";
+}
+
+static std::string GenerateCodeCache(
+    std::map<std::string, ScriptCompiler::CachedData*> data) {
+  std::stringstream ss;
+  ss << R"(// This file is generated by tools/mkcodecache
+
+#include "node_native_module.h"
+#include "v8.h"
+#include "util.h"
+
+#include <cinttypes>
+
+namespace node {
+namespace native_module {
+)";
+
+  size_t total = 0;
+  for (const auto& x : data) {
+    const std::string& id = x.first;
+    ScriptCompiler::CachedData* cached_data = x.second;
+    total += cached_data->length;
+    std::string def = GetDefinition(id, cached_data->length, cached_data->data);
+    ss << def << "\n\n";
+
+    std::cout << "Generated cache for " << id
```

Comment:
I tried this out locally, and noticed that the code cache generator is run multiple times during a `make test` - this is not a urgent issue (for instance, we run the v8 default mksnapshot multiple times as well), but we may want to silence this output (we could make it conditional with `credentials::SafeGetenv('NODE_DEBUG', &text) && text == "mkcodecache"` or search for `-v` in argv)

Reply:
I added the environment variable switch in https://github.com/nodejs/node/pull/27135 (just noticed that there is uv_os_getenv() now)

Reply:
Rebased this whole PR on https://github.com/nodejs/node/pull/27135 verbatim

Reply:
P.S. this is a special CLI tool. I can take a `--verbose` CLI arg.